mesonbuild/wrap/wrap.py

```diff
@@ -383,18 +384,44 @@ def download(self, what: str, ofname: str) -> None:
         os.rename(tmpfile, ofname)
 
     def get_file_internal(self, what: str) -> str:
-        filename = self.wrap.get(what + '_filename')
-        cache_path = os.path.join(self.cachedir, filename)
+        if all([x in self.wrap.values for x in [what + '_url', what + '_path']]):
+            raise WrapException('Both {}_url and {}_path were specified'.format(what, what))
 
-        if os.path.exists(cache_path):
-            self.check_hash(what, cache_path)
-            mlog.log('Using', mlog.bold(self.packagename), what, 'from cache.')
+        if what + '_url' in self.wrap.values:
+            filename = self.wrap.get(what + '_filename')
+            cache_path = os.path.join(self.cachedir, filename)
+
+            if os.path.exists(cache_path):
+                self.check_hash(what, cache_path)
+                mlog.log('Using', mlog.bold(self.packagename), what, 'from cache.')
+                return cache_path
+
+            if not os.path.isdir(self.cachedir):
+                os.mkdir(self.cachedir)
+            self.download(what, cache_path)
             return cache_path
+        elif what + '_path' in self.wrap.values:
+            from ..interpreterbase import FeatureNew
+            FeatureNew('Local wrap patch files via {}_path'.format(what), '0.54.0').use('')
+            path = Path(self.wrap.get(what + '_path'))
+
+            # check that the path is within subprojects and exists
+            if path.is_absolute():
+                raise WrapException('{}_path ({}) must be relative to the subprojects directory'.format(what, path))
+            path = Path(self.subdir_root) / path
+
+            if not path.exists():
+                raise WrapException('File "{}" does not exist'.format(path))
 
-        if not os.path.isdir(self.cachedir):
-            os.mkdir(self.cachedir)
-        self.download(what, cache_path)
-        return cache_path
+            path = path.resolve()
+            try:
+                path.relative_to(Path(self.subdir_root))
+            except Exception:
+                raise WrapException('{}_path must be within the root subprojects directory'.format(what))
+
+            return path.as_posix()
```

Comment:
We should call self.check_hash() if we have `what + '_hash' in self.wrap.values`, just in case.